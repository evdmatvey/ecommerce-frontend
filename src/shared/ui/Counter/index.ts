export { default } from './ui/Counter';
