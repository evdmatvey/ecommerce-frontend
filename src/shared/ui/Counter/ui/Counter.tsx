import clsx from 'clsx';
import { MinusIcon, PlusIcon } from '../../Icons';
import styles from './Counter.module.css';

interface CounterProps {
  count: number;
  min?: number;
  max?: number;
  size?: 'small' | 'medium' | 'large';
  decrement: () => void;
  increment: () => void;
}

const Counter = ({
  size = 'medium',
  count,
  min,
  max,
  decrement,
  increment,
}: CounterProps) => {
  const classes = clsx(styles.root, styles[size]);

  return (
    <div className={classes}>
      <button
        className={styles.decrement}
        onClick={decrement}
        disabled={min ? count === min : false}
        data-testid="decrement"
      >
        <MinusIcon />
      </button>
      <span className={styles.count}>{count}</span>
      <button
        className={styles.increment}
        onClick={increment}
        disabled={max ? count === max : false}
        data-testid="increment"
      >
        <PlusIcon />
      </button>
    </div>
  );
};

export default Counter;
