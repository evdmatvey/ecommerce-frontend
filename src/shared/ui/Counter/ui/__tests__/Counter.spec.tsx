import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import Counter from '../Counter';

describe('Counter logic', () => {
  it('should render the component with the default size', () => {
    render(<Counter count={0} decrement={() => {}} increment={() => {}} />);
    expect(screen.getByTestId('decrement')).toBeInTheDocument();
    expect(screen.getByTestId('increment')).toBeInTheDocument();
    expect(screen.getByText('0')).toBeInTheDocument();
    expect(screen.getByTestId('decrement')).not.toBeDisabled();
    expect(screen.getByTestId('increment')).not.toBeDisabled();
  });

  it('should call the decrement function when the decrement button is clicked', () => {
    const decrementMock = vi.fn();
    render(
      <Counter count={0} decrement={decrementMock} increment={() => {}} />,
    );
    fireEvent.click(screen.getByTestId('decrement'));
    expect(decrementMock).toHaveBeenCalledTimes(1);
  });

  it('should call the increment function when the increment button is clicked', () => {
    const incrementMock = vi.fn();
    render(
      <Counter count={0} decrement={() => {}} increment={incrementMock} />,
    );
    fireEvent.click(screen.getByTestId('increment'));
    expect(incrementMock).toHaveBeenCalledTimes(1);
  });

  it("shouldn't call the increment function when the increment button is clicked", () => {
    const incrementMock = vi.fn();
    render(
      <Counter
        count={1}
        max={1}
        decrement={() => {}}
        increment={incrementMock}
      />,
    );
    fireEvent.click(screen.getByTestId('increment'));
    expect(incrementMock).not.toHaveBeenCalledTimes(1);
  });

  it('should disable the decrement button when the count is at the minimum', () => {
    render(
      <Counter count={1} min={1} decrement={() => {}} increment={() => {}} />,
    );
    expect(screen.getByTestId('decrement')).toBeDisabled();
    expect(screen.getByTestId('increment')).not.toBeDisabled();
  });

  it('should disable the increment button when the count is at the maximum', () => {
    render(
      <Counter count={10} max={10} decrement={() => {}} increment={() => {}} />,
    );
    expect(screen.getByTestId('increment')).toBeDisabled();
    expect(screen.getByTestId('decrement')).not.toBeDisabled();
  });
});

describe('Counter snapshots', () => {
  it('Counter with disabled decrement', () => {
    const { asFragment } = render(
      <Counter
        count={1}
        min={1}
        max={10}
        increment={() => {}}
        decrement={() => {}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('Counter with disabled increment', () => {
    const { asFragment } = render(
      <Counter
        count={1}
        min={10}
        max={10}
        increment={() => {}}
        decrement={() => {}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
