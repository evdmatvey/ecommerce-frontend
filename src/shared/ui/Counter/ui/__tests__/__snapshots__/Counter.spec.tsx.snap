// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Counter snapshots > Counter with disabled decrement 1`] = `
<DocumentFragment>
  <div
    class="_root_0163fa _medium_0163fa"
  >
    <button
      class="_decrement_0163fa"
      data-testid="decrement"
      disabled=""
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.125 12C4.125 11.5168 4.51675 11.125 5 11.125H19C19.4832 11.125 19.875 11.5168 19.875 12C19.875 12.4832 19.4832 12.875 19 12.875H5C4.51675 12.875 4.125 12.4832 4.125 12Z"
          fill="#181D25"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <span
      class="_count_0163fa"
    >
      1
    </span>
    <button
      class="_increment_0163fa"
      data-testid="increment"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12 4.125C12.4832 4.125 12.875 4.51675 12.875 5V19C12.875 19.4832 12.4832 19.875 12 19.875C11.5168 19.875 11.125 19.4832 11.125 19V5C11.125 4.51675 11.5168 4.125 12 4.125Z"
          fill="#181D25"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M4.125 12C4.125 11.5168 4.51675 11.125 5 11.125H19C19.4832 11.125 19.875 11.5168 19.875 12C19.875 12.4832 19.4832 12.875 19 12.875H5C4.51675 12.875 4.125 12.4832 4.125 12Z"
          fill="#181D25"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Counter snapshots > Counter with disabled increment 1`] = `
<DocumentFragment>
  <div
    class="_root_0163fa _medium_0163fa"
  >
    <button
      class="_decrement_0163fa"
      data-testid="decrement"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.125 12C4.125 11.5168 4.51675 11.125 5 11.125H19C19.4832 11.125 19.875 11.5168 19.875 12C19.875 12.4832 19.4832 12.875 19 12.875H5C4.51675 12.875 4.125 12.4832 4.125 12Z"
          fill="#181D25"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <span
      class="_count_0163fa"
    >
      1
    </span>
    <button
      class="_increment_0163fa"
      data-testid="increment"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12 4.125C12.4832 4.125 12.875 4.51675 12.875 5V19C12.875 19.4832 12.4832 19.875 12 19.875C11.5168 19.875 11.125 19.4832 11.125 19V5C11.125 4.51675 11.5168 4.125 12 4.125Z"
          fill="#181D25"
          fill-rule="evenodd"
        />
        <path
          clip-rule="evenodd"
          d="M4.125 12C4.125 11.5168 4.51675 11.125 5 11.125H19C19.4832 11.125 19.875 11.5168 19.875 12C19.875 12.4832 19.4832 12.875 19 12.875H5C4.51675 12.875 4.125 12.4832 4.125 12Z"
          fill="#181D25"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
